source/server/listener_manager_impl.cc

```diff
@@ -694,20 +706,49 @@ void ListenerManagerImpl::onListenerWarmed(ListenerImpl& listener) {
   updateWarmingActiveGauges();
 }
 
-void ListenerManagerImpl::drainFilterChains(ListenerImplPtr&& listener) {
+void ListenerManagerImpl::inPlaceFilterChainUpdate(ListenerImpl& listener) {
+  auto existing_active_listener = getListenerByName(active_listeners_, listener.name());
+  auto existing_warming_listener = getListenerByName(warming_listeners_, listener.name());
+  ASSERT(existing_warming_listener != warming_listeners_.end());
+  ASSERT(*existing_warming_listener != nullptr);
+
+  (*existing_warming_listener)->debugLog("execute in place filter chain update");
+  if (existing_active_listener != active_listeners_.end()) {
+    ASSERT(*existing_active_listener != nullptr);
+    // The warmed listener should be added first so that the worker will accept new connections
+    // when it stops listening on the old listener.
+    for (const auto& worker : workers_) {
+      addListenerToWorker(*worker, listener.listenerTag(), listener, nullptr);
+    }
+    // Finish active_listeners_ transformation before calling `drainListener` as it depends on their
+    // state.
+    auto previous_listener = std::move(*existing_active_listener);
+    *existing_active_listener = std::move(*existing_warming_listener);
+    drainFilterChains(std::move(previous_listener), **existing_active_listener);
+  } else {
+    for (const auto& worker : workers_) {
```

Comment:
If we are doing an in place update? How does this else branch ever happen? Why wouldn't there be an active listener? More comments please.

Reply:
I don't think it could happen either... Will add comment and asssert

Reply:
Add the comment that stop/remove should cancel the in place warming listener.